### DISCUSSION TITLE
Hub to Hub - Fernbedienung LED Umschaltung

### DISCUSSION THREAD

**Discussion by msw-home** [pybricks/support#1746] [Q&A]
Moin zusammen, ich bin es schon wieder.
Ich habe einen gelben und einen grünen Spike Prime Hub.
Und die Lego Fernbedienung, 6 Steuertasten, 1 Center Taste und die LED Anzeige.
Ich möchte die LED Anzeige bei jedem Tastendruck der Center Taste, eine farbliche Änderung erreichen.
weiß - grün - gelb - rot und wieder von vorn.
Funktioniert mit einem Hub ohne Probleme.

Jetzt möchte ich mittels eines 2 Hub`s die gleichen Funktionen realisieren.
Ich verstehe die Hub to Hub Logik immer noch nicht.
Ich habe verschiedene Varianten ausprobiert, ohne Erfolg.
Wo ist/sind meine Denkfehler?
In welchen Hub müssen welche Befehle?
Ist die Kommunikation der beiden Hub`s nicht korrekt?
Hilfe wäre wünschenswert. 
Anbei die beiden Programme der beiden Hub`s.
![gelber_Hub---pybricks_program_2024-08-07T11_00_29](https://github.com/user-attachments/assets/af1b7220-ca83-4de3-a02c-d66a787039df)
![grüen_Hub---pybricks_program_2024-08-07T11_00_40](https://github.com/user-attachments/assets/15c22b9a-554b-47b6-a777-16ba6cc336d7)

**Comment by BertLindeman** [comment]
Manfred,

It is a bit much to read and understand the whole programs. 
What I understand from looking at the programs:

yellow hub
- broadcast ch 5
- observe ch 6
- connects to remote control

green hub
- broadcast ch 6
- observe ch 5

I see the yellow hub continuous sending and receiving the button status.
Wouldn't it be enough just to send the button status (or something coded) and the status-color from Yellow to Green?
Then observe on Green and act upon them. 
It's hard (for me) to see and really understand the programs on these prints.
Although  they are way better than screenprints.  

 Bert

**Reply by msw-home** [reply]
Guten Morgen Bert,
ich kann verstehen, dass das alles sehr unübersichtlich ist.
Mir würde ja im 1. Ansatz schon genügen, wie eine Hub to Hub Kopplung auszusehen hat.
Mir einer Erklärung wie der Datenaustausch abläuft.

Hub 1 sendet die 7 Tasten der Remote an Hub 2, wie lautet der Blockbefehl?
Hub 2 empfängt die 7 Daten, wie lautet der Blockbefehl?
Hub 2 sendet vom Colorsensor die Farben an Hub 1, wie lautet der Blockbefehl?
Anschlüsse am Hub 1. 6 Motoren
Anschlüsse an Hub2, 2 Motoren, 1 Colorsensor, 1 Lichtmatrix 
An welchen Hub 1 oder 2, gehört die Remote?
Welche Parameter gehören wo hin?

VG Manfred

![pybricks_program_2024-08-08T05_50_20](https://github.com/user-attachments/assets/ed3c4432-2089-441b-aaed-0309e61f47ab)
![pybricks_program_2024-08-08T05_53_03](https://github.com/user-attachments/assets/719207ae-a553-41fc-affb-7552eac0f7f2)

**Reply by BertLindeman** [reply]
Manfred,

In no way am I a solid programmer, but this is getting more clear to me.
And I hope you do not want a complete program, but want to get to know the hub to hub communication a bit better.

Lets keep the names at yellow_hub and green_hub (in any language 😄 )

An answer I can give up front:
>  An welchen Hub 1 oder 2, gehört die Remote?
> Welche Parameter gehören wo hin?

In your previous programs the yellow hub connects to the remote control.
Let us keep it that way, why not.
Each hub does its own programming, and if it needs variables for that, those variables are defines for that hub.
(Not so clear, sorry)

Do you want me to **try** to write two programs that use hub-to-hub to only do the color-change loop?
So you see what it takes to do that and can build upon that.
Or implement in your own program.  

Bert

**Reply by msw-home** [reply]
Hallo Bert, vorab schon mal danke für alles.

Das wäre klasse, sowas als Vorlage zu bekommen.
Soll ich versuchen , zwei Programme zu schreiben, die Hub-to-Hub verwenden, um nur die Farbwechselschleife auszuführen?
Damit Sie sehen, was dazu nötig ist, und darauf aufbauen können.

Gruß Manfred

**Reply by BertLindeman** [reply]
Manfred,

It will take a bit of time, as I am not used to do the block-coding. 
So an opportunity for me too.
I am basing the program upon [Laurens project](https://pybricks.com/project/technic-42146-powered-up-remote/)
So you might have a peek there?

Bert

**Comment by BertLindeman** [comment]
Manfred,

A yellow_hub program. 
It is so clumsy (according to me) but it loops through the four status colors.
It transmits a list with 
- status
- button-left
- button-left-plus
- button-left-minus
- button-right
- button-right-plus
- button-right-minus

[manfred_yellow.py.txt](https://github.com/user-attachments/files/16547982/manfred_yellow.py.txt)
You can download it, rename to remove the txt suffix and import in pybricks.

As implied: no testing on the receive side done yet.

Bert

**Reply by BertLindeman** [reply]
Manfred,

Updated yellow program (did not fill status each color change) and the very simple green program:

[manfred_green.py.txt](https://github.com/user-attachments/files/16548544/manfred_green.py.txt)
[manfred_yellow.py.txt](https://github.com/user-attachments/files/16548550/manfred_yellow.py.txt)

Remove .txt and import in pybricks.

Time to get outside now. The temperature is up to it.

Bert

**Reply by BertLindeman** [reply]
Manfred,

Hope you did not used too much time on my previous programs (full of bugs).

Have structured the programs a bit more to get them a bit more clear.

On the green hub the received buttons are shown on the matrix display, two digits.
The left digit for the left buttons and (surprise) on the right the right buttons:
- 1 plus button
- 2 red button
- 4 minus button
So each side can show 0 to 7 for the combinations.

If you encounter firmware level problems, let me know. My hubs are not always on the stable version.

My coding time for today is up. 
Work in the local library is next.

Current programs:
[manfred_green_V01.py.txt](https://github.com/user-attachments/files/16559479/manfred_green_V01.py.txt)
[manfred_yellow_V01.py.txt](https://github.com/user-attachments/files/16559482/manfred_yellow_V01.py.txt)

Curious if these programs look huge to you, as your programs looked huge to me in the start.

I will try to keep these programs as they are, unless I encounter more bugs. 

Enjoy,

Bert

**Reply by msw-home** [reply]
Hallo Bert,
das sieht erstmal ganz super aus!!!!
Beim ersten kurzen Versuch hat alles geklappt.
Ja, die Programme sind für mich riesig.
Um sie zu verstehen, muss ich wohl erst ein Informatikstudium absolvieren!
Wenn ich das meinem 8 jährigen Enkel, als Kinderspielzeug verkaufen will, wird er (und seine Eltern), mich für verrückt erklären.
Aber jeder braucht neue Herausforderungen.
Danke, für alles!

Bleib gesund, wir bleiben in Kontakt.

Manfred


Hello Bert,
that looks great at first!!!!
Everything worked on the first short attempt.
Yes, the programs are huge for me.
To understand them, I'll probably have to study computer science first!
If I try to sell this to my 8-year-old grandson as a toy, he (and his parents) will think I'm crazy.
But everyone needs new challenges.
Thanks for everything!

Stay healthy, we'll stay in touch.

Manfred

**Reply by BertLindeman** [reply]
> Wenn ich das meinem 8 jährigen Enkel, als Kinderspielzeug verkaufen will, wird er (und seine Eltern), mich für verrückt erklären.

Dass ihr alle zusammen Spaß habt, das ist wichtig.

Bert

**Reply by msw-home** [reply]
Hallo Bert,

da bin ich schon wieder.
Habe jetzt deine Programme und meine Programmteile zusammengeführt.
Läuft alles wie es soll! Juhu!!!

Jetzt fehlt mir noch 1 Parameter,
ich habe einen Color Sensor an den green Hub angeschlossen, funktioniert auch alles.
Jetzt brauche ich aber noch die Meldung "Farbe erkannt" 
vom green Hub zum yellow Hub übertragen, und weiß ich nicht weiter!

Anbei meine zusammen geführten Programm.
[V03_green.py.txt](https://github.com/user-attachments/files/16565031/V03_green.py.txt)
[V03_yellow.py.txt](https://github.com/user-attachments/files/16565032/V03_yellow.py.txt)

**Reply by BertLindeman** [reply]
Manfred, 

Thanks for including the text programs, so I can look at them locally in stead of on a picture.

So,  just a broadcast of True or False would do the trick?
You might do a broadcast on green at the moment you know if the color **is** recognized.
I think you know 😄 

And on the location in the yellow program, where you need to know the color is recognized, do the observe.

I will take a look probably tomorrow. 

Have a nice weekend,

Bert

**[EDIT]**

 It is too hot to sit inside programming.
It will take more time for me to really look at it.

**Reply by msw-home** [reply]
Hallo Bert, kein Zeitdruck vorhanden, ich schaue mir das auch erst nächste Woche wieder an.
Wochenende geht vor!
Ich möchte es so realisieren
rot erkannt = 2 übertragen
gelb erkannt = 4 übertragen
blau erkannt = 6 übertragen
alles andere = gleich nicht erkannt = 99 übertragen

**Reply by BertLindeman** [reply]
'n-abend Manfred.

Before it is getting really hot tomorrow, I added a small part to the green program:
- extra variable `Farbe_fur_gelbe_hub` value to transmit to the yellow hub
- set `Farbe_fur_gelbe_hub` like you stated: 2 / 4 / 6 or 99
- broadcast the value.

In the python side the function now it looks like:

```python

async def subtask3():
    global Farbe_erkannt, Farbe_fur_gelbe_hub
    # rot erkannt = 2 übertragen
    # gelb erkannt = 4 übertragen
    # blau erkannt = 6 übertragen
    # alles andere = gleich nicht erkannt = 99 übertragen
    while True:
        await wait(1)
        if await color_sensor.color() == Color.RED:
            Farbe_erkannt = 1
            Farbe_fur_gelbe_hub = 2 * Farbe_erkannt
        elif await color_sensor.color() == Color.YELLOW:
            Farbe_erkannt = 2
            Farbe_fur_gelbe_hub = 2 * Farbe_erkannt
        elif await color_sensor.color() == Color.BLUE:
            Farbe_erkannt = 3
            Farbe_fur_gelbe_hub = 2 * Farbe_erkannt
        elif await color_sensor.color() == Color.NONE:
            Farbe_erkannt = 0
            Farbe_fur_gelbe_hub = 99
        else:
            Farbe_fur_gelbe_hub = 99
        await green_hub.ble.broadcast([Farbe_fur_gelbe_hub])
```
The green program (made it version 4)
[V04_green.py.txt](https://github.com/user-attachments/files/16575831/V04_green.py.txt)
 
For the yellow hub I do not know what you want to use the `Farbe_fur_gelbe_hub` for.
So I declare that homework for you, 
define a variable and determine what the value is to be used for. 
And find the correct spot in the yellow program to add that.
And add the observe.

As example I used variable `Farbe_code_green_hub`
The observe might look like this:
![afbeelding](https://github.com/user-attachments/assets/e6fd4fd4-8dd7-4be0-aba6-47da1208a65a)

Hope you get a step further with this.
I think I stated this before: I can not test these programs as I do not have all devices you attach.

Bert

**Reply by msw-home** [reply]
Hallo Bert,
läuft alles so wie es soll, ihr (Du) wart mir eine große Hilfe.
Alle Bewegungen funktionieren im Einrichtbetrieb.
Werde jetzt meine Lego Mechanik auf den gewünschten Stand bringen.
Danke nochmals!
Gruß Manfred

**Reply by BertLindeman** [reply]
Wow Manfred

Really cool (in these temperatures).

If you want to show it off here, would be really appreciated. 

Thank you for the status report.

Bert
